restapi/src/main/java/io/stargate/web/docsapi/service/DocsSchemaChecker.java

```diff
@@ -0,0 +1,53 @@
+package io.stargate.web.docsapi.service;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import io.stargate.db.schema.Schema;
+import io.stargate.web.docsapi.dao.DocumentDB;
+import io.stargate.web.docsapi.exception.ErrorCode;
+import io.stargate.web.docsapi.exception.ErrorCodeRuntimeException;
+import io.stargate.web.docsapi.service.util.ImmutableKeyspaceAndTable;
+
+/**
+ * A helper class for checking that particular keyspace/table combinations exist and are valid
+ * document API collections.
+ */
+public class DocsSchemaChecker {
+  private Schema lastCheckedSchema;
+
+  private final Cache<ImmutableKeyspaceAndTable, Boolean> validatedDocCollections =
+      Caffeine.newBuilder().maximumSize(10_000).build();
+
+  private void clearCacheOnSchemaChange(DocumentDB db) {
+    if (!db.schema().equals(lastCheckedSchema)) {
+      validatedDocCollections.invalidateAll();
+    }
+  }
+
+  public void checkValidity(String keyspace, String table, DocumentDB db) {
+    ImmutableKeyspaceAndTable keyspaceAndTable =
+        ImmutableKeyspaceAndTable.builder().keyspace(keyspace).table(table).build();
+    clearCacheOnSchemaChange(db);
+    Boolean valid = validatedDocCollections.getIfPresent(keyspaceAndTable);
+    if (valid == null) {
+      // no cache hit
+      performValidityCheck(keyspaceAndTable, db);
+    }
+  }
+
+  private void performValidityCheck(ImmutableKeyspaceAndTable keyspaceAndTable, DocumentDB db) {
+    String keyspace = keyspaceAndTable.getKeyspace();
+    String table = keyspaceAndTable.getTable();
+    db.tableExists(keyspace, table);
+    boolean isDocsTable = db.isDocumentsTable(keyspace, table);
+    if (!isDocsTable) {
+      throw new ErrorCodeRuntimeException(
+          ErrorCode.DOCS_API_TABLE_NOT_A_COLLECTION,
+          String.format(
+              "The Cassandra table %s.%s is not a Documents collection. Accessing arbitrary tables via the Documents API is not permitted.",
+              keyspace, table));
+    }
+    validatedDocCollections.put(keyspaceAndTable, true);
+    this.lastCheckedSchema = db.schema();
```

Comment:
Would it be more appropriate to do this assignment in `clearCacheOnSchemaChange()`?